File: src/JavaScript/devpossible-ton/src/formatter/TonFormatter.ts
/**
 * TonFormatter - Code formatter for TON format
 * Copyright (c) 2024 DevPossible, LLC
 */

import { TonParser } from '../parser/TonParser';
import { TonSerializer } from '../serializer/TonSerializer';
import { TonSerializeOptions, TonFormatStyle } from '../serializer/TonSerializeOptions';

export interface TonFormatterOptions {
  style?: TonFormatStyle;
  indentSize?: number;
  indentChar?: string;
  sortProperties?: boolean;
  preserveComments?: boolean;
  trailingCommas?: boolean;
  quoteStyle?: 'single' | 'double';
  lineEnding?: string;
  maxLineLength?: number;
}

export class TonFormatter {
  private options: TonFormatterOptions;

  constructor(options?: TonFormatterOptions) {
    this.options = {
      style: TonFormatStyle.Pretty,
      indentSize: 4,
      indentChar: ' ',
      sortProperties: false,
      preserveComments: true,
      trailingCommas: false,
      quoteStyle: 'single',
      lineEnding: '\n',
      maxLineLength: 80,
      ...options
    };
  }

  /**
   * Formats TON content
   */
  public format(content: string): string {
    try {
      // Parse the content
      const parser = new TonParser();
      const document = parser.parse(content);

      // If we need to preserve comments, we need a different approach
      if (this.options.preserveComments) {
        return this.formatWithComments(content);
      }

      // Use the serializer to format
      const serializeOptions: TonSerializeOptions = {
        formatStyle: this.options.style,
        indentSize: this.options.indentSize,
        indentChar: this.options.indentChar,
        sortProperties: this.options.sortProperties,
        trailingCommas: this.options.trailingCommas,
        quoteStyle: this.options.quoteStyle,
        lineEnding: this.options.lineEnding,
        includeHeader: true,
        includeTypeHints: true,
        omitNulls: false,
        omitUndefined: false
      };

      const serializer = new TonSerializer(serializeOptions);
      return serializer.serialize(document);
    } catch (error) {
      // If parsing fails, return original content
      console.error('Failed to format TON content:', error);
      return content;
    }
  }

  /**
   * Formats TON content while preserving comments
   */
  private formatWithComments(content: string): string {
    // This is a simplified version - a full implementation would need
    // to track comment positions and reinsert them appropriately
    const lines = content.split(/\r?\n/);
    const formatted: string[] = [];
    let depth = 0;

    for (const line of lines) {
      const trimmed = line.trim();

      // Skip empty lines
      if (!trimmed) {
        if (this.options.style === TonFormatStyle.Pretty) {
          formatted.push('');
        }
        continue;
      }

      // Handle comments
      if (trimmed.startsWith('//')) {
        formatted.push(this.getIndent(depth) + trimmed);
        continue;
      }

      // Handle block comments
      if (trimmed.startsWith('/*')) {
        formatted.push(this.getIndent(depth) + trimmed);
        continue;
      }

      // Adjust depth for braces
      const openBraces = (trimmed.match(/\{/g) || []).length;
      const closeBraces = (trimmed.match(/\}/g) || []).length;
      const openBrackets = (trimmed.match(/\[/g) || []).length;
      const closeBrackets = (trimmed.match(/\]/g) || []).length;

      const depthBefore = depth;
      depth += openBraces + openBrackets;
      depth -= closeBraces + closeBrackets;

      // Use appropriate indentation
      const currentIndent = closeBraces > 0 || closeBrackets > 0 ?
        Math.min(depthBefore, depth) : depthBefore;

      if (this.options.style === TonFormatStyle.Pretty) {
        formatted.push(this.getIndent(currentIndent) + trimmed);
      } else {
        // Compact style - join lines
        if (formatted.length > 0 && !trimmed.startsWith('#')) {
          const last = formatted[formatted.length - 1];
          if (!last.endsWith('{') && !last.endsWith('[')) {
            formatted[formatted.length - 1] = last + ' ' + trimmed;
          } else {
            formatted[formatted.length - 1] = last + trimmed;
          }
        } else {
          formatted.push(trimmed);
        }
      }
    }

    return formatted.join(this.options.lineEnding);
  }

  /**
   * Gets indentation string for a given depth
   */
  private getIndent(depth: number): string {
    if (this.options.style === TonFormatStyle.Compact) {
      return '';
    }
    return (this.options.indentChar || ' ').repeat((this.options.indentSize || 4) * depth);
  }

  /**
   * Validates that the formatted content is equivalent to original
   */
  public validate(original: string, formatted: string): boolean {
    try {
      const parser = new TonParser();
      const originalDoc = parser.parse(original);
      const formattedDoc = parser.parse(formatted);

      // Compare the JSON representations
      const originalJson = JSON.stringify(originalDoc.toJSON());
      const formattedJson = JSON.stringify(formattedDoc.toJSON());

      return originalJson === formattedJson;
    } catch (error) {
      return false;
    }
  }
}

/**
 * Static helper functions
 */
export namespace TonFormatter {
  /**
   * Format TON content with default pretty style
   */
  export function pretty(content: string): string {
    const formatter = new TonFormatter({ style: TonFormatStyle.Pretty });
    return formatter.format(content);
  }

  /**
   * Format TON content with compact style
   */
  export function compact(content: string): string {
    const formatter = new TonFormatter({ style: TonFormatStyle.Compact });
    return formatter.format(content);
  }

  /**
   * Format and sort properties alphabetically
   */
  export function sorted(content: string): string {
    const formatter = new TonFormatter({
      style: TonFormatStyle.Pretty,
      sortProperties: true
    });
    return formatter.format(content);
  }
}

File: src/Python/devpossible_ton/devpossible_ton/formatter/ton_formatter.py
"""
TonFormatter - Code formatter for TON format
Copyright (c) 2024 DevPossible, LLC
"""

from typing import Optional
from ..parser import TonParser, TonParseOptions
from ..serializer import TonSerializer
from ..serializer.ton_serialize_options import TonSerializeOptions, TonFormatStyle


class TonFormatter:
    """Formatter for TON format"""
    
    def __init__(self, options: Optional[TonSerializeOptions] = None):
        self.options = options or TonSerializeOptions.pretty()
    
    def format(self, content: str) -> str:
        """Formats TON content"""
        parser = TonParser()
        document = parser.parse(content)
        
        serializer = TonSerializer(self.options)
        return serializer.serialize(document)
    
    def validate(self, original: str, formatted: str) -> bool:
        """Validates that the formatted content is equivalent to original"""
        try:
            parser = TonParser()
            original_doc = parser.parse(original)
            formatted_doc = parser.parse(formatted)
            
            # Compare the JSON representations
            import json
            original_json = json.dumps(original_doc.to_json(), sort_keys=True)
            formatted_json = json.dumps(formatted_doc.to_json(), sort_keys=True)
            
            return original_json == formatted_json
        except Exception:
            return False
    
    @staticmethod
    def format_string(content: str, style: TonFormatStyle = TonFormatStyle.PRETTY) -> str:
        """Format TON content with specified style"""
        if style == TonFormatStyle.COMPACT:
            options = TonSerializeOptions.compact()
        else:
            options = TonSerializeOptions.pretty()
        
        formatter = TonFormatter(options)
        return formatter.format(content)
    
    @staticmethod
    async def format_string_async(content: str, style: TonFormatStyle = TonFormatStyle.PRETTY) -> str:
        """Format TON content with specified style asynchronously"""
        import asyncio
        return await asyncio.to_thread(TonFormatter.format_string, content, style)
    
    @staticmethod
    def format_file(file_path: str, style: TonFormatStyle = TonFormatStyle.PRETTY) -> str:
        """Formats a TON file using the specified style"""
        if not file_path:
            raise ValueError("file_path cannot be None or empty")
        
        import os
        if not os.path.exists(file_path):
            raise FileNotFoundError(f"File not found: {file_path}")
        
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        return TonFormatter.format_string(content, style)
    
    @staticmethod
    async def format_file_async(file_path: str, style: TonFormatStyle = TonFormatStyle.PRETTY) -> str:
        """Formats a TON file using the specified style asynchronously"""
        import aiofiles
        import os
        
        if not file_path:
            raise ValueError("file_path cannot be None or empty")
        
        if not os.path.exists(file_path):
            raise FileNotFoundError(f"File not found: {file_path}")
        
        async with aiofiles.open(file_path, 'r', encoding='utf-8') as f:
            content = await f.read()
        
        return await TonFormatter.format_string_async(content, style)
    
    @staticmethod
    def format_file_in_place(file_path: str, style: TonFormatStyle = TonFormatStyle.PRETTY) -> None:
        """Formats a TON file in place using the specified style"""
        formatted_content = TonFormatter.format_file(file_path, style)
        
        with open(file_path, 'w', encoding='utf-8') as f:
            f.write(formatted_content)
    
    @staticmethod
    async def format_file_in_place_async(file_path: str, style: TonFormatStyle = TonFormatStyle.PRETTY) -> None:
        """Formats a TON file in place using the specified style asynchronously"""
        import aiofiles
        
        formatted_content = await TonFormatter.format_file_async(file_path, style)
        
        async with aiofiles.open(file_path, 'w', encoding='utf-8') as f:
            await f.write(formatted_content)
    
    @staticmethod
    def format_file_to_file(input_file_path: str, output_file_path: str, 
                           style: TonFormatStyle = TonFormatStyle.PRETTY) -> None:
        """Formats TON content to a new file using the specified style"""
        if not input_file_path:
            raise ValueError("input_file_path cannot be None or empty")
        if not output_file_path:
            raise ValueError("output_file_path cannot be None or empty")
        
        formatted_content = TonFormatter.format_file(input_file_path, style)
        
        with open(output_file_path, 'w', encoding='utf-8') as f:
            f.write(formatted_content)
    
    @staticmethod
    async def format_file_to_file_async(input_file_path: str, output_file_path: str,
                                       style: TonFormatStyle = TonFormatStyle.PRETTY) -> None:
        """Formats TON content to a new file using the specified style asynchronously"""
        import aiofiles
        
        if not input_file_path:
            raise ValueError("input_file_path cannot be None or empty")
        if not output_file_path:
            raise ValueError("output_file_path cannot be None or empty")
        
        formatted_content = await TonFormatter.format_file_async(input_file_path, style)
        
        async with aiofiles.open(output_file_path, 'w', encoding='utf-8') as f:
            await f.write(formatted_content)
    
    @staticmethod
    def pretty(content: str) -> str:
        """Format TON content with pretty style"""
        return TonFormatter.format_string(content, TonFormatStyle.PRETTY)
    
    @staticmethod
    def compact(content: str) -> str:
        """Format TON content with compact style"""
        return TonFormatter.format_string(content, TonFormatStyle.COMPACT)
    
    @staticmethod
    def sorted(content: str) -> str:
        """Format and sort properties alphabetically"""
        options = TonSerializeOptions.pretty()
        options.sort_properties = True
        
        formatter = TonFormatter(options)
        return formatter.format(content)
